File: C#/SubmodularFunction/Submodular/Manual.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

#if DEBUG
namespace Onigiri.Submodular
{
    public class Manual:SubmodularOracle
    {
        double[] values;

        public Manual(int n,double []array)
        {
            SetVariable(n, array[0]);
            values = array;;
        }

        
        internal override double Value(int[] order, int cardinality)
        {
            int mask = 0;
            for (int i = 0; i < cardinality; i++)
            {
                mask |= 1 << order[i];
            }//for i
            return values[mask];
        }

        public double Value(int mask)
        {
            return values[mask];
        }

        //protected override void ContractDerived(int element)
        //{
        //    throw new NotImilementedException();
        //}

    }
}

#endif

File: C++/SubmodularFunctionOld/SubmodularFunction/Manual.cpp
#ifdef _DEBUG

#include "Submodular.h"

using namespace OnigiriSubmodular;

Manual::Manual(int n,const double* array)
{
	Manual::n = n;
	Manual::values = new double[1<<n];
	for(int i=0;i<n;i++){
		Manual::values[i] = array[i];
	}
	Manual::fOfEmpty = values[0];
}

Manual::~Manual(){
	delete[] Manual::values;
}

double Manual::Value(const int* order,int cardinality)
{
	int mask = 0;
	for (int i = 0; i < cardinality; i++)
	{
		mask |= 1 << order[i];
	}//for i
	return values[mask];
}

double Manual:: Value(int mask)
{
	return values[mask];
}

void Manual::Copy(const Manual &other){
	Manual::n = other.n;
	Manual::values = new double[n];
	for(int i=0;i<n;i++){
		values[i] = other.values[i];
	}
}

#endif